src/lib/y2packager/product_finder.rb

```diff
@@ -0,0 +1,139 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) 2019 SUSE LLC, All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+# ------------------------------------------------------------------------------
+
+require "solv"
+
+module Y2Packager
+  # This class finds products in a Solv pool
+  class ProductFinder
+    #
+    # Constructor
+    #
+    # @param pool [Solv::Pool] the pool used for evaluating the products
+    #
+    def initialize(pool)
+      @pool = pool.pool
+    end
+
+    #
+    # Evaluate all products in the pool and return details about them
+    # including the dependencies.
+    #
+    # @param base [String] The name of the base product used for evaluating the
+    #  dependencies.
+    #
+    # @return [Array<Hash>] The list of found products
+    #
+    def products(base = nil)
+      marked_base_products = base_product_tags
+      # evaluate all products
+      pool.whatprovides(pool.str2id("product()")).each_with_object([]) do |s, list|
```

Comment:
this block is a bit too big. I suggest to simply have something like
```
pool.whatprovides(pool.str2id("product()")).each_with_object([]) do |s, list|
  list.concat(create_products(s)
end
```

btw that "product()" looks like string.